src/lib.rs

```diff
@@ -303,4 +277,53 @@ fn restore_extended_key(
     })
 }
 
+struct TxBuilder {
+    recipients: Vec<(String, u64)>,
+    fee_rate: Option<f32>,
+}
+
+impl TxBuilder {
+    fn new() -> Self {
+        TxBuilder {
+            recipients: Vec::new(),
+            fee_rate: None,
+        }
+    }
+
+    fn add_recipient(&self, recipient: String, amount: u64) -> Arc<Self> {
+        let mut recipients = self.recipients.to_vec();
+        recipients.append(&mut vec![(recipient, amount)]);
+        Arc::new(TxBuilder {
+            recipients,
+            fee_rate: self.fee_rate,
+        })
+    }
+
+    fn fee_rate(&self, sat_per_vb: f32) -> Arc<Self> {
+        Arc::new(TxBuilder {
+            recipients: self.recipients.to_vec(),
+            fee_rate: Some(sat_per_vb),
+        })
+    }
+
+    fn build(&self, wallet: &Wallet) -> Result<Arc<PartiallySignedBitcoinTransaction>, Error> {
+        let wallet = wallet.get_wallet();
+        let mut tx_builder = wallet.build_tx();
+        for (address, amount) in &self.recipients {
+            let address = Address::from_str(address).expect("recipient address parsing failed");
```

Comment:
Can we move this to the `add_recipient` function so validation is done there (with possible error Result)?

Reply:
We can, but let's talk about this today. Having a result chain breaks the builder pattern. We either want to fail at every step or expose a builder. Having failing intermediate steps is not usually how a builder is exposed.

Reply:
I did a little more reading on the builder pattern and I think you're right that it's more user friendly to chain setters together without having to worry about any input validation errors until the build end. You can disregard my suggestion. :slightly_smiling_face: